scenario/state.py

```diff
@@ -341,6 +421,39 @@ class Relation(_RelationBase):
         default_factory=lambda: {0: DEFAULT_JUJU_DATABAG.copy()},  # dedup
     )
 
+    # NOTE: This can be replaced with dataclasses.KW_ONLY in Python 3.10+
+    def __init__(
+        self,
+        endpoint: str,
+        interface: Optional[str] = None,
+        *,
+        id: Optional[int] = None,
+        local_app_data: Optional["RawDataBagContents"] = None,
+        local_unit_data: Optional["RawDataBagContents"] = None,
+        remote_app_name: str = "remote",
+        limit: int = 1,
+        remote_app_data: Optional["RawDataBagContents"] = None,
+        remote_units_data: Optional[Dict["UnitID", "RawDataBagContents"]] = None,
+    ):
+        object.__setattr__(self, "remote_app_name", remote_app_name)
+        object.__setattr__(self, "limit", limit)
+        object.__setattr__(self, "remote_app_data", remote_app_data or {})
+        object.__setattr__(
+            self,
+            "remote_units_data",
+            {0: DEFAULT_JUJU_DATABAG.copy()}
+            if remote_units_data is None
+            else remote_units_data,
+        )
+        # Do the super init last, as it will call _validate_databag on the databags.
```

Comment:
Kind of unrelated, but this makes me think: can we put the databag validation step in the consistency checker?
Can't recall precisely why I wanted that done so early on.

Reply:
and I know pydantic is a big dependency, but we could also consider using pydantic for State. Would make sense to be a bit stricter with the types in here rather than having to manually check things later

Reply:
Let's have this discussion a different time (kind of unrelated as you said).